NAS-135866 / 25.10 / Remove unwanted newlines in API docs

creatorcary opened the pull request:
The following field docstring:
```python
    field: int
    """Just a
    chill guy."""
```
produces the field description `"Just a\nchill guy."` which isn't pleasant to look at in the API docs.

## Rejected Solution
Parse double newlines as single newlines and single newlines like in the example above as spaces.

Using this implicit docstring parsing makes some formatting unintuitive or unwieldy, particularly when using reST in API docstrings.

## Solution
Instead, we need to be intentional when writing API docstrings by ending line wraps with backslashes to remove the newline. So the above example would become:
```python
    field: int
    """Just a \
    chill guy."""
```
and would parse to `"Just a chill guy."` as intended.

## Proposed Standards

Now that our API docs are [provided to the public](https://api.truenas.com/v25.10.0/), we need to establish some formatting guidelines. The changes made in this PR include:
* Only use triple quotation marks `"""` for API docstrings as per [PEP-257](https://peps.python.org/pep-0257/#:~:text=For%20consistency%2C%20always%20use%20%22%22%22triple%20double%20quotes%22%22%22). Triple apostrophes `'''`, single quotation marks `"`, and single apostrophes `'` should not be used to denote docstrings.
* API docstrings should begin with a capital letter unless the first word is a variable name or other identifier.
* API docstrings all must end with a period.

bugclerk commented:
Jira URL: https://ixsystems.atlassian.net/browse/NAS-135866

themylogin commented on src/middlewared/middlewared/api/base/jsonschema.py:
```diff
@@ -3,25 +3,45 @@
+    return _clean_field_descriptions(_replace_refs(schema, defs))
+
+
+def _replace_refs(data, defs=None):
```
Why was this method name changed?

creatorcary replied:
Everywhere where it is called we also want to call `_clean_field_schemas` so I combined them into one utility function that can be imported instead of importing both. The idea is that we no longer need to export this function, unless there is a case in which we would want to replace refs without cleaning descriptions?

themylogin replied:
If you don't need to export a function, don't put it to `__all__`, why the underscore names?

creatorcary replied:
`__all__` is only used to replace the `*` in `from module import *`. It still allows `from module import replace_refs`.

I usually try to honor https://peps.python.org/pep-0008/#:~:text=_single_leading_underscore%3A%20weak%20%E2%80%9Cinternal%20use%E2%80%9D%20indicator.%20E.g.%20from%20M%20import%20*%20does%20not%20import%20objects%20whose%20names%20start%20with%20an%20underscore. when I can because it tells future developers that they shouldn't import the function in other modules (in this case, they should instead import `clean_schema`).

Given this argument, I'm still willing to change if we don't like the leading underscores.

themylogin commented on src/middlewared/middlewared/api/base/jsonschema.py:
```diff
@@ -3,25 +3,45 @@
+
+    :param field_schema: `model.model_json_schema()`
+    """
+    for field_schema in schema["properties"].values():
```
How are descriptions cleaned on nested fields?

creatorcary replied:
Thanks for the catch

creatorcary replied:
Will complete after https://ixsystems.atlassian.net/browse/NAS-135959

creatorcary commented:
The API docs are publicly facing now, so better care will just have to be taken. Here's a concrete example of where the original solution is not ideal.

Here's an unordered list in RST normally:
```
* this is
* a list

  * with a nested list
  * and some subitems

* and here the parent list continues
```
If each double newline is parsed into a single newline, here's how we would have to write it in the docstring:
```python
"""
* this is
* a list



  * with a nested list
  * and some subitems



* and here the parent list continues
"""
```

creatorcary commented:
* Added a new flake8 plugin `flake8-literal` which can be used to standardize string literals across the repository. For now, just using it to enforce the proposed standard of using `"""` for docstrings. It also adds checks for raw and formatted strings.
* Added a CI test to loosely check capitalization, punctuation, and line wraps in API docstrings. It's neither ideal nor optimal, but it will help enforce standards moving forward.
* Fixed up the docstrings that have been added/changed since this PR was created.

bugclerk commented:
This PR has been merged and conversations have been locked.
If you would like to discuss more about this issue please use our forums or raise a Jira ticket.
